File: sdks/csharp/src/AgentBrowser.cs
using System.Diagnostics;
using System.Runtime.InteropServices;

namespace B4N1Web;

/// <summary>
/// B4n1Web Agent Browser
/// 
/// A browser instance optimized for AI agent workflows.
/// Requires B4n1Web binary to be installed.
/// </summary>
/// <example>
/// <code>
/// using B4N1Web;
/// 
/// var browser = new AgentBrowser(new BrowserOptions { Mode = BrowserMode.Light });
/// var page = await browser.GotoAsync("https://example.com");
/// Console.WriteLine(page.Markdown);
/// browser.Close();
/// </code>
/// </example>
public class AgentBrowser : IDisposable
{
    private const string SdkVersion = "0.12.0";
    private readonly BrowserOptions _options;
    private readonly string _binaryPath;
    private readonly IProcessRunner _runner;
    private Page? _lastPage;
    private bool _disposed;

    public AgentBrowser(BrowserOptions? options = null) : this(options, new RealProcessRunner())
    {
    }

    public AgentBrowser(BrowserOptions? options, IProcessRunner runner)
    {
        _options = options ?? new BrowserOptions();
        _runner = runner ?? new RealProcessRunner();

        _binaryPath = FindBinary();
        if (string.IsNullOrEmpty(_binaryPath))
        {
            throw new BinaryNotFoundException();
        }

    }

    /// <summary>
    /// Navigate to a URL and extract structured content
    /// </summary>
    public async Task<Page> GotoAsync(string url, string? waitFor = null)
    {
        var args = $"goto {url} --mode {_options.Mode.ToString().ToLower()}";
        if (waitFor != null)
        {
            args += $" --wait-for \"{waitFor}\"";
        }

        var startInfo = CreateStartInfo(args);
        var result = await _runner.RunAsync(startInfo, _options.Timeout * 1000);

        if (result.TimedOut)
        {
            throw new Exception($"Binary timed out after {_options.Timeout}s");
        }

        if (result.ExitCode != 0)
        {
            throw new Exception($"Binary error: {result.StdErr}");
        }

        var page = ParseOutput(url, result.StdOut);
        _lastPage = page;
        return page;
    }

    /// <summary>
    /// Navigate to URL (synchronous version)
    /// </summary>
    public Page Goto(string url, string? waitFor = null)
    {
        return GotoAsync(url, waitFor).GetAwaiter().GetResult();
    }

    /// <summary>
    /// Take a screenshot of the current page
    /// </summary>
    public string Screenshot(int width, int height)
    {
        if (_lastPage == null)
            throw new InvalidOperationException("No page loaded. Call Goto or GotoAsync first.");

        var startInfo = CreateStartInfo($"screenshot --url {_lastPage.Url} --width {width} --height {height}");
        var result = _runner.Run(startInfo, _options.Timeout * 1000);

        if (result.TimedOut)
        {
            throw new Exception($"Binary timed out after {_options.Timeout}s");
        }

        if (result.ExitCode != 0)
        {
            throw new Exception($"Binary error: {result.StdErr}");
        }

        foreach (var line in result.StdOut.Split('\n'))
        {
            if (line.StartsWith("Screenshot:"))
            {
                var b64 = line[11..].Trim();
                if (!string.IsNullOrEmpty(b64))
                    return b64;
            }
        }

        throw new Exception("No screenshot data returned from binary");
    }

    /// <summary>
    /// Wait for a CSS selector to appear on the page
    /// </summary>
    public bool WaitForSelector(string selector, int timeoutMs)
    {
        if (_lastPage == null)
            throw new InvalidOperationException("No page loaded. Call Goto or GotoAsync first.");

        var startInfo = CreateStartInfo($"wait-for-selector \"{selector}\" --url {_lastPage.Url} --timeout {timeoutMs}");
        var result = _runner.Run(startInfo, _options.Timeout * 1000);

        if (result.TimedOut)
        {
            throw new Exception($"Binary timed out after {_options.Timeout}s");
        }

        if (result.ExitCode != 0)
        {
            throw new Exception($"Binary error: {result.StdErr}");
        }

        foreach (var line in result.StdOut.Split('\n'))
        {
            if (line.StartsWith("Found:"))
            {
                return line[6..].Trim() == "true";
            }
        }

        return false;
    }

    /// <summary>
    /// Click on an element by CSS selector
    /// </summary>
    public void Click(string selector)
    {
        if (_lastPage == null)
            throw new InvalidOperationException("No page loaded. Call Goto or GotoAsync first.");

        var startInfo = CreateStartInfo($"click \"{selector}\" --url {_lastPage.Url}");
        var result = _runner.Run(startInfo, _options.Timeout * 1000);

        if (result.TimedOut)
        {
            throw new Exception($"Binary timed out after {_options.Timeout}s");
        }

        if (result.ExitCode != 0)
        {
            throw new Exception($"Binary error: {result.StdErr}");
        }
    }

    /// <summary>
    /// Type text into an element by CSS selector
    /// </summary>
    public void TypeText(string selector, string text, bool clearFirst)
    {
        if (_lastPage == null)
            throw new InvalidOperationException("No page loaded. Call Goto or GotoAsync first.");

        var clearArg = clearFirst ? " --clear-first" : "";
        var startInfo = CreateStartInfo($"type-text \"{selector}\" \"{text}\" --url {_lastPage.Url}{clearArg}");
        var result = _runner.Run(startInfo, _options.Timeout * 1000);

        if (result.TimedOut)
        {
            throw new Exception($"Binary timed out after {_options.Timeout}s");
        }

        if (result.ExitCode != 0)
        {
            throw new Exception($"Binary error: {result.StdErr}");
        }
    }

    /// <summary>
    /// Get links from the last visited page
    /// </summary>
    public string[] GetLinks()
    {
        return _lastPage?.Links.ToArray() ?? Array.Empty<string>();
    }

    /// <summary>
    /// Fetch links from a URL without creating a browser instance (static)
    /// </summary>
    public static string[] GetLinksFromPage(string url, IProcessRunner? runner = null)
    {
        var binaryPath = FindBinary();
        if (string.IsNullOrEmpty(binaryPath))
            throw new BinaryNotFoundException();

        runner ??= new RealProcessRunner();
        var startInfo = new ProcessStartInfo
        {
            FileName = binaryPath,
            Arguments = $"goto {url} --mode light",
            UseShellExecute = false,
            RedirectStandardOutput = true,
            RedirectStandardError = true,
            CreateNoWindow = true
        };

        var result = runner.Run(startInfo, 30000);

        if (result.ExitCode != 0)
            return Array.Empty<string>();

        foreach (var line in result.StdOut.Split('\n'))
        {
            if (line.StartsWith("Links:"))
            {
                try
                {
                    return System.Text.Json.JsonSerializer.Deserialize<List<string>>(line[6..])
                        ?.ToArray() ?? Array.Empty<string>();
                }
                catch
                {
                    return Array.Empty<string>();
                }
            }
        }

        return Array.Empty<string>();
    }

    /// <summary>
    /// Parse text output from the binary
    /// </summary>
    private Page ParseOutput(string url, string output)
    {
        var markdown = new List<string>();
        var links = new List<string>();
        string? screenshot = null;
        string? jsOutput = null;

        foreach (var line in output.Split('\n'))
        {
            if (line.StartsWith("URL:"))
            {
                continue;
            }
            else if (line.StartsWith("Markdown:"))
            {
                continue;
            }
            else if (line.StartsWith("Links:"))
            {
                try
                {
                    links = System.Text.Json.JsonSerializer.Deserialize<List<string>>(line[6..]) 
                        ?? new List<string>();
                }
                catch
                {
                    links = new List<string>();
                }
            }
            else if (line.StartsWith("Screenshot:"))
            {
                screenshot = line[11..].Trim();
                if (string.IsNullOrEmpty(screenshot)) screenshot = null;
            }
            else if (line.StartsWith("JsOutput:"))
            {
                jsOutput = line[9..].Trim();
                if (string.IsNullOrEmpty(jsOutput)) jsOutput = null;
            }
            else
            {
                markdown.Add(line);
            }
        }

        return new Page
        {
            Url = url,
            Markdown = string.Join("\n", markdown).Trim(),
            Links = links,
            Screenshot = screenshot,
            JsOutput = jsOutput
        };
    }

    /// <summary>
    /// Find b4n1web binary in bundled location or system install
    /// </summary>
    private static string FindBinary()
    {
        // 1. Check bundled binary (extracted from embedded resource for current platform)
        var bundledPath = ExtractBundledBinary();
        if (!string.IsNullOrEmpty(bundledPath))
        {
            return bundledPath;
        }

        // 2. Check system install locations
        var possiblePaths = new[]
        {
            "/usr/local/bin/b4n1web",
            "/usr/bin/b4n1web",
            Path.Combine(Environment.GetFolderPath(Environment.SpecialFolder.UserProfile), ".local/bin/b4n1web"),
            Path.Combine(Environment.GetFolderPath(Environment.SpecialFolder.UserProfile), ".b4n1web/bin/b4n1web"),
        };

        foreach (var path in possiblePaths)
        {
            if (File.Exists(path))
            {
                return path;
            }
        }

        return string.Empty;
    }

    /// <summary>
    /// Get the platform-specific embedded resource name for the current OS/arch
    /// </summary>
    private static string? GetPlatformResourceName()
    {
        string os;
        if (RuntimeInformation.IsOSPlatform(OSPlatform.Linux))
            os = "linux";
        else if (RuntimeInformation.IsOSPlatform(OSPlatform.OSX))
            os = "macos";
        else if (RuntimeInformation.IsOSPlatform(OSPlatform.Windows))
            os = "windows";
        else
            return null;

        string arch = RuntimeInformation.OSArchitecture switch
        {
            Architecture.X64 => "amd64",
            Architecture.Arm64 => "arm64",
            _ => null!
        };
        if (arch == null) return null;

        string ext = os == "windows" ? ".exe" : "";
        return $"B4N1Web.native.{os}-{arch}.b4n1web{ext}";
    }

    /// <summary>
    /// Extract bundled binary for current platform from embedded resources to temp directory
    /// </summary>
    private static string? ExtractBundledBinary()
    {
        var resourceName = GetPlatformResourceName();
        if (resourceName == null) return null;

        string ext = RuntimeInformation.IsOSPlatform(OSPlatform.Windows) ? ".exe" : "";
        var tempBinary = Path.Combine(Path.GetTempPath(), "b4n1web", $"b4n1web{ext}");

        if (File.Exists(tempBinary))
        {
            try
            {
                var psi = new ProcessStartInfo
                {
                    FileName = tempBinary,
                    Arguments = "--version",
                    UseShellExecute = false,
                    RedirectStandardOutput = true,
                    RedirectStandardError = true,
                    CreateNoWindow = true
                };
                using var proc = new Process { StartInfo = psi };
                proc.Start();
                var output = proc.StandardOutput.ReadToEnd().Trim();
                proc.WaitForExit(3000);
                if (proc.ExitCode == 0 && !string.IsNullOrEmpty(output))
                    return tempBinary;
            }
            catch
            {
            }
        }

        try
        {
            var assembly = typeof(AgentBrowser).Assembly;
            using var stream = assembly.GetManifestResourceStream(resourceName);
            if (stream == null) return null;

            var tempDir = Path.Combine(Path.GetTempPath(), "b4n1web");
            Directory.CreateDirectory(tempDir);

            using var fs = new FileStream(tempBinary, FileMode.Create, FileAccess.Write);
            stream.CopyTo(fs);

            // Make executable on Unix
            if (RuntimeInformation.IsOSPlatform(OSPlatform.Linux) ||
                RuntimeInformation.IsOSPlatform(OSPlatform.OSX))
            {
                var chmod = new Process
                {
                    StartInfo = new ProcessStartInfo("chmod", $"+x {tempBinary}")
                    {
                        UseShellExecute = false
                    }
                };
                chmod.Start();
                chmod.WaitForExit();
            }

            return tempBinary;
        }
        catch
        {
            return null;
        }
    }

    /// <summary>
    /// Get B4n1Web binary version
    /// </summary>
    public static string GetVersion(IProcessRunner? runner = null)
    {
        var path = FindBinary();
        if (string.IsNullOrEmpty(path))
        {
            return "unknown";
        }

        try
        {
            runner ??= new RealProcessRunner();
            var startInfo = new ProcessStartInfo
            {
                FileName = path,
                Arguments = "--version",
                UseShellExecute = false,
                RedirectStandardOutput = true,
                CreateNoWindow = true
            };

            var result = runner.Run(startInfo, 5000);
            return result.StdOut.Trim();
        }
        catch
        {
            return "unknown";
        }
    }

    /// <summary>
    /// Close the browser session
    /// </summary>
    public void Close()
    {
        Dispose();
    }

    public void Dispose()
    {
        if (!_disposed)
        {
            _disposed = true;
        }
        GC.SuppressFinalize(this);
    }

    private ProcessStartInfo CreateStartInfo(string arguments)
    {
        return new ProcessStartInfo
        {
            FileName = _binaryPath,
            Arguments = arguments,
            UseShellExecute = false,
            RedirectStandardOutput = true,
            RedirectStandardError = true,
            CreateNoWindow = true
        };
    }
}

File: sdks/java/src/main/java/com/b4n1/web/AgentBrowser.java
package com.b4n1.web;

import java.io.BufferedReader;
import java.io.File;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.List;
import java.util.Locale;
import java.util.Objects;
import java.util.stream.Collectors;

/**
 * B4n1Web Agent Browser.
 *
 * A browser instance optimized for AI agent workflows.
 * Requires B4n1Web binary to be installed.
 *
 * <pre>{@code
 * var browser = new AgentBrowser(BrowserMode.LIGHT);
 * var page = browser.goto("https://example.com");
 * System.out.println(page.getMarkdown());
 * browser.close();
 * }</pre>
 */
public class AgentBrowser implements AutoCloseable {
    private static final String SDK_VERSION = "0.13.0";
    private final BrowserOptions options;
    private final String binaryPath;

    public AgentBrowser() {
        this(new BrowserOptions());
    }

    public AgentBrowser(BrowserOptions options) {
        this.options = Objects.requireNonNull(options, "options cannot be null");

        String path = findBinary();
        if (path == null || path.isEmpty()) {
            throw new BinaryNotFoundException();
        }
        this.binaryPath = path;
    }

    AgentBrowser(BrowserOptions options, String testBinaryPath) {
        this.options = Objects.requireNonNull(options, "options cannot be null");
        if (testBinaryPath != null) {
            this.binaryPath = testBinaryPath;
        } else {
            String path = findBinary();
            if (path == null || path.isEmpty()) {
                throw new BinaryNotFoundException();
            }
            this.binaryPath = path;
        }
    }

    /**
     * Navigate to a URL and extract structured content.
     */
    public Page goto_(String url) throws NavigationException {
        return goto_(url, null);
    }

    /**
     * Navigate to a URL with an optional CSS selector to wait for.
     */
    public Page goto_(String url, String waitFor) throws NavigationException {
        try {
            List<String> cmd = new ArrayList<>();
            cmd.add(binaryPath);
            cmd.add("goto");
            cmd.add(url);
            cmd.add("--mode");
            cmd.add(options.getMode().getValue());
            if (waitFor != null && !waitFor.isEmpty()) {
                cmd.add("--wait-for");
                cmd.add(waitFor);
            }

            ProcessBuilder pb = new ProcessBuilder(cmd);
            pb.redirectErrorStream(true);
            Process process = pb.start();

            String output = new BufferedReader(new InputStreamReader(process.getInputStream()))
                    .lines()
                    .collect(Collectors.joining("\n"));

            boolean finished = process.waitFor(options.getTimeout(), java.util.concurrent.TimeUnit.SECONDS);
            if (!finished) {
                process.destroyForcibly();
                throw new NavigationException(url, "Binary timed out after " + options.getTimeout() + "s");
            }

            if (process.exitValue() != 0) {
                throw new NavigationException(url, "Binary error: exit code " + process.exitValue());
            }

            return parseOutput(url, output);

        } catch (NavigationException e) {
            throw e;
        } catch (Exception e) {
            throw new NavigationException(url, e);
        }
    }

    /**
     * Take a screenshot of the current page with specified dimensions.
     */
    public String screenshot(int width, int height) {
        try {
            ProcessBuilder pb = new ProcessBuilder(
                binaryPath, "screenshot", "--width", String.valueOf(width), "--height", String.valueOf(height),
                "--mode", options.getMode().getValue()
            );
            pb.redirectErrorStream(true);
            Process process = pb.start();

            String output = new BufferedReader(new InputStreamReader(process.getInputStream()))
                    .lines()
                    .collect(Collectors.joining("\n"));

            process.waitFor(options.getTimeout(), java.util.concurrent.TimeUnit.SECONDS);
            return output.trim();
        } catch (Exception e) {
            throw new RuntimeException("Screenshot failed", e);
        }
    }

    /**
     * Wait for a CSS selector to appear within the given timeout.
     */
    public boolean waitForSelector(String selector, int timeoutMs) {
        try {
            ProcessBuilder pb = new ProcessBuilder(
                binaryPath, "wait-selector", selector, "--timeout", String.valueOf(timeoutMs),
                "--mode", options.getMode().getValue()
            );
            pb.redirectErrorStream(true);
            Process process = pb.start();

            String output = new BufferedReader(new InputStreamReader(process.getInputStream()))
                    .lines()
                    .collect(Collectors.joining("\n"));

            process.waitFor(timeoutMs / 1000 + 1, java.util.concurrent.TimeUnit.SECONDS);
            return process.exitValue() == 0;
        } catch (Exception e) {
            return false;
        }
    }

    /**
     * Click an element matching the given CSS selector.
     */
    public void click(String selector) {
        try {
            ProcessBuilder pb = new ProcessBuilder(
                binaryPath, "click", selector, "--mode", options.getMode().getValue()
            );
            pb.redirectErrorStream(true);
            Process process = pb.start();
            process.waitFor(options.getTimeout(), java.util.concurrent.TimeUnit.SECONDS);
        } catch (Exception e) {
            throw new RuntimeException("Click failed for selector: " + selector, e);
        }
    }

    /**
     * Type text into an element matching the given CSS selector.
     */
    public void typeText(String selector, String text, boolean clearFirst) {
        try {
            List<String> cmd = new ArrayList<>();
            cmd.add(binaryPath);
            cmd.add("type");
            cmd.add(selector);
            cmd.add(text);
            cmd.add("--mode");
            cmd.add(options.getMode().getValue());
            if (clearFirst) {
                cmd.add("--clear-first");
            }

            ProcessBuilder pb = new ProcessBuilder(cmd);
            pb.redirectErrorStream(true);
            Process process = pb.start();
            process.waitFor(options.getTimeout(), java.util.concurrent.TimeUnit.SECONDS);
        } catch (Exception e) {
            throw new RuntimeException("TypeText failed for selector: " + selector, e);
        }
    }

    private Page parseOutput(String url, String output) {
        StringBuilder markdown = new StringBuilder();
        List<String> links = new ArrayList<>();
        String jsOutput = null;

        for (String line : output.split("\n")) {
            if (line.startsWith("URL:")) {
                continue;
            } else if (line.startsWith("Markdown:")) {
                continue;
            } else if (line.startsWith("Links:")) {
                String linksJson = line.substring(6).trim();
                links = parseLinksJson(linksJson);
            } else if (line.startsWith("JSOutput:")) {
                jsOutput = line.substring(9).trim();
            } else {
                if (markdown.length() > 0) {
                    markdown.append("\n");
                }
                markdown.append(line);
            }
        }

        Page page = new Page();
        page.setUrl(url);
        page.setMarkdown(markdown.toString().trim());
        page.setLinks(links);
        page.setJsOutput(jsOutput);
        return page;
    }

    private List<String> parseLinksJson(String json) {
        if (json == null || json.isEmpty() || json.equals("[]")) {
            return new ArrayList<>();
        }
        json = json.trim();
        if (!json.startsWith("[")) {
            return new ArrayList<>();
        }
        List<String> result = new ArrayList<>();
        json = json.substring(1, json.length() - 1);
        for (String item : json.split(",")) {
            item = item.trim();
            if (item.startsWith("'") && item.endsWith("'")) {
                result.add(item.substring(1, item.length() - 1));
            } else if (item.startsWith("\"") && item.endsWith("\"")) {
                result.add(item.substring(1, item.length() - 1));
            }
        }
        return result;
    }

    private String findBinary() {
        try {
            String resourcePath = getPlatformResourcePath();
            if (resourcePath != null) {
                java.net.URL url = getClass().getResource(resourcePath);
                if (url != null) {
                    File bundledBinary = extractBundledBinary(resourcePath);
                    if (bundledBinary != null && bundledBinary.canExecute()) {
                        return bundledBinary.getAbsolutePath();
                    }
                }
            }
        } catch (Exception e) {
        }

        String[] possiblePaths = {
            "/usr/local/bin/b4n1web",
            "/usr/bin/b4n1web",
            System.getProperty("user.home") + "/.local/bin/b4n1web",
            System.getProperty("user.home") + "/.b4n1web/bin/b4n1web"
        };

        for (String path : possiblePaths) {
            File f = new File(path);
            if (f.exists() && f.canExecute()) {
                return path;
            }
        }
        return null;
    }

    private static String getPlatformResourcePath() {
        String os = System.getProperty("os.name").toLowerCase(Locale.ROOT);
        String arch = System.getProperty("os.arch").toLowerCase(Locale.ROOT);

        String osPart;
        if (os.contains("linux")) {
            osPart = "linux";
        } else if (os.contains("mac") || os.contains("darwin")) {
            osPart = "macos";
        } else if (os.contains("windows")) {
            osPart = "windows";
        } else {
            return null;
        }

        String archPart;
        if (arch.equals("amd64") || arch.equals("x86_64")) {
            archPart = "amd64";
        } else if (arch.equals("aarch64") || arch.equals("arm64")) {
            archPart = "arm64";
        } else {
            return null;
        }

        String ext = os.contains("windows") ? ".exe" : "";
        return "/native/" + osPart + "-" + archPart + "/b4n1web" + ext;
    }

    private File extractBundledBinary(String resourcePath) {
        try {
            String ext = resourcePath.endsWith(".exe") ? ".exe" : "";
            File tempDir = new File(System.getProperty("java.io.tmpdir"), "b4n1web");
            tempDir.mkdirs();
            File tempBinary = new File(tempDir, "b4n1web" + ext);

            try (java.io.InputStream in = getClass().getResourceAsStream(resourcePath);
                 java.io.FileOutputStream out = new java.io.FileOutputStream(tempBinary)) {
                if (in == null) return null;
                byte[] buffer = new byte[8192];
                int bytesRead;
                while ((bytesRead = in.read(buffer)) != -1) {
                    out.write(buffer, 0, bytesRead);
                }
            }
            tempBinary.setExecutable(true);
            return tempBinary;
        } catch (Exception e) {
            return null;
        }
    }

    /**
     * Get B4n1Web binary version.
     */
    public static String getVersion() {
        String[] paths = {
            "/usr/local/bin/b4n1web",
            "/usr/bin/b4n1web",
            System.getProperty("user.home") + "/.local/bin/b4n1web",
            System.getProperty("user.home") + "/.b4n1web/bin/b4n1web"
        };

        for (String path : paths) {
            File f = new File(path);
            if (f.exists() && f.canExecute()) {
                try {
                    Process pb = new ProcessBuilder(path, "--version").start();
                    String version = new BufferedReader(new InputStreamReader(pb.getInputStream()))
                            .lines()
                            .collect(Collectors.joining("\n"));
                    pb.waitFor(5, java.util.concurrent.TimeUnit.SECONDS);
                    String trimmed = version.trim();
                    String[] parts = trimmed.split(" ");
                    return parts.length >= 2 ? parts[parts.length - 1] : trimmed;
                } catch (Exception e) {
                }
            }
        }
        return "unknown";
    }

    @Override
    public void close() {
    }
}
